H29549 on h29610

wangkx opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-29610 Fix broken file spray when no DropZoneRestriction

1. When spraying a file with no DropZoneRestriction, the
getAndValidateDropZone() may not return a dropzone plane. The
readAndCheckSpraySourceReq() should validate the return and read the
plane name if there is the return.
2. In the checkDZScopeAccessAndSetSpraySourceDFUFileSpec(), validate
DZPathScopePermissions only if the dropzone plane name is found in
the readAndCheckSpraySourceReq().
3. Create throwOrLogDropZoneLookUpError(): if not bare metal or
DropZoneRestriction enabled, throw an exception; otherwise, log the
error.
4. Create findDropZonePlaneName(): if a plane name is not found based
on dropzone host and path, call throwOrLogDropZoneLookUpError().
5. In ESP FileSpray service, call the findDropZonePlaneName() to find
a dropzone plane name. Validate DZPathScopePermissions only if the
plane name is found.
6. In getDZPathScopePermissions(), if a dropzone plane is not found
based on dropzone host and path, call throwOrLogDropZoneLookUpError().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx committed:
HPCC-29549 Handle fully qualified path when validating DZ scope access

1. Move the code of the scope permission validation from the
checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to the
readAndCheckSpraySourceReq().
2. In readAndCheckSpraySourceReq(), parse the file requests inside the
SourcePath request. For each file request, validating the scope permission.
If a file request contains a host/ip, validate the host/ip and local file
path. If a file request does not contain a host/ip, validate the file path.
3. Rename the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to
setSpraySourceDFUFileSpec().
4. Validate localhost in getDropZoneHost().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1916,38 +1960,95 @@ void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char*
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            if (plane)
-                sourcePlaneReq.append(plane->queryProp("@name"));
+            validateHostReqInSprayPathReq(sourceIPReq, sourcePathReq); //The sourceIPReq is from the original request.
```
related to comment on line 1987, validateHostReqInSprayPathReq will be iterate through files again.

There are 3 separate coded iterations through 'files' in this function (1 in validateHostReqInSprayPathReq ), it feels like a single loop that dealt with dropzone, non-dropzone and UNC paths would be clearer .

wangkx replied:
In https://github.com/hpcc-systems/HPCC-Platform/pull/17356/commits/e51b2e9e539c3589cd2f4099ee2422df548b1664, I tried doing them in a single loop. But, it was not easy to read (understand the logic). I think the 3 loops here are better for read. For one spray, only 2 of them are called for either a plane is found or not found. In both cases, the loop around line 1987 will be called and the loop should not cost much.

jakesmith replied:
ok, well as long as all paths have been tested, this code will be refactored soon anyway.

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1916,37 +1960,95 @@ void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char*
-            sourcePlaneReq.append(srcPlane);
         }
-        else
+        if (sourcePlaneReq.isEmpty()) // must be true, unless bare-metal and isDropZoneRestrictionEnabled()==false
```
My mistake. Will change.

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1916,37 +1960,95 @@ void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char*
+
+            //Validate file path.
+            StringBuffer s;
+            if (isAbsolutePath(path))
```
Will rename the isHostInSprayPath to isUNCPath. 
The isUNCPath does imply an absolute path. The isAbsolutePath can be under the else branch. It was not in there to not duplicate the code for isPathInPlane. I will move isAbsolutePath under the else branch.

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1916,38 +1960,95 @@ void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char*
+                const char* file = files.item(i);
+                if (!isEmptyString(file))
+                {
+                    path.set(file);
```
I think that the isPathInPlane call later will stop a request where 1st is a path under dropzone, but rest aren't. Do I miss anything?

jakesmith replied:
Then there should be a comment somewhere, perhaps after line 1995, to say that all other paths must be in this dropzone.

wangkx replied:
The comment is added.

wangkx committed:
HPCC-29549 Revise based on review

Fix 1 comment, rename 1 call, and only check absolute path if not UNC
Add one comment about the path validation for all files.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review my changes.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2014,29 +2016,34 @@ void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, MemoryBuffer
             {
+                path = localPath.str();
                 //Based on the tests, the dfuserver only supports the wildcard inside the file name, like '/path/f*'.
                 //The dfuserver throws an error if the wildcard is inside the path, like /p*ath/file.
```
I think this comment is now misplaced, it was related to the ScopePermissions and tailing wildcard for a File lookup (which looks at leading scopes). This comment should be moved.

wangkx replied:
I will move this comment back to the place the getDZFileScopePermissions is called. It is here because someone may be wondering what if a wildcard is inside the path and the isPathInPlane returns a false.

jakesmith replied:
>. It is here because someone may be wondering what if a wildcard is inside the path and the isPathInPlane returns a false.

suggests needs different comment here also.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2014,29 +2016,34 @@ void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, MemoryBuffer
+                path = localPath.str();
                 //Based on the tests, the dfuserver only supports the wildcard inside the file name, like '/path/f*'.
                 //The dfuserver throws an error if the wildcard is inside the path, like /p*ath/file.
                 if (!isPathInPlane(dropZone, path))
```
It would make more sense if validateUNCPath also did this check.. it's doing other checks and it has the dropZone

wangkx replied:
Yes. Renaming to validateUNCPath makes more sense if it also does this check.

jakesmith replied:
ok yes, although it wasn't just doing "is UNC Path" before this proposed change.

jakesmith reviewed (commented):
@wangkx - please see new comments.

wangkx committed:
HPCC-29549 Revise based on review

1. Rename isUNCPath to validateUNCPath. Revise it and
related code.
2. Move a comment.
3. Add one comment
4. Always check the hostInPath with the hostEp

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1928,7 +1928,11 @@ static bool isUNCPath(const char* sprayPath, const char* dropZoneName, IProperty
         }
     }
     else if (!isHostInPlane(dropZone, hostInPath, true))
```
only calling isHostInPlane for UNC (when dropzone) - when sourceIPReq is not supplied?

In dropzone case, if sourceIPReq is supplied, what checks the localPath of the UNC is within it?

wangkx replied:
I think this block (and the if (!hostEp.isNull())) should be removed because the hostEp is always set.  If the sourcePlaneReq is empty, the sourceIPReq cannot be empty. If the sourcePlaneReq is not empty, the getDropZoneHost will be called with a possible sourceIPReq. In getDropZoneHost, if the sourceIPReq is empty, the sourceIPReq will be filled in. if the sourceIPReq is not empty, the sourceIPReq will be validated by isHostInPlane. Agree?

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1917,18 +1917,18 @@ static bool isUNCPath(const char* sprayPath, const char* dropZoneName, IProperty
-            msg.append("A dropzone name specified in the SourcePlane is preferred. The host/IP should not be contained in the SourcePath.");
-            throw makeStringException(ECLWATCH_INVALID_INPUT, msg.str());
-        }
+        VStringBuffer msg("The host %s defined in the SourcePath does not match with the host %s defined in SourceIP. ", hostInPath.str(), host);
```
I can remove 'defined in SourceIP'. How about the next line '"A dropzone name specified in the SourcePlane is preferred.'? When throwing this exception, the SourcePlane may already be specified. Should I check the dropZone before adding that sentence?

wangkx commented:
A new PR will be created.
